File: backend/models/Geofence.js
const mongoose = require('mongoose');

const geofenceSchema = mongoose.Schema(
  {
    name: {
      type: String,
      required: true,
    },
    address: {
      type: String,
      default: '',
    },
    latitude: {
      type: Number,
      required: true,
    },
    longitude: {
      type: Number,
      required: true,
    },
    radius: {
      type: Number,
      required: true,
    },
    shape: {
      type: String,
      enum: ['circle', 'polygon'],
      default: 'circle',
    },
    isActive: {
      type: Boolean,
      default: true,
    },
    createdBy: {
      type: String,
    },
    assignedEmployees: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
    ],
    // A–Z labeling for geofences
    labelLetter: {
      type: String,
      // Optional to maintain backward compatibility; UI will require on new geofences
    },
  },
  { timestamps: true }
);

// Enforce unique label letter per active geofence
geofenceSchema.index(
  { labelLetter: 1 },
  { unique: true, partialFilterExpression: { isActive: true, labelLetter: { $exists: true } } }
);

module.exports = mongoose.model('Geofence', geofenceSchema);

File: field_check/lib/models/geofence.dart
class Geofence {
  final String id;
  final String name;
  final double latitude;
  final double longitude;
  final double radius;

  Geofence({
    required this.id,
    required this.name,
    required this.latitude,
    required this.longitude,
    required this.radius,
  });

  factory Geofence.fromJson(Map<String, dynamic> json) {
    return Geofence(
      id: json['_id'] as String,
      name: json['name'] as String,
      latitude: (json['coordinates'][0] as num).toDouble(),
      longitude: (json['coordinates'][1] as num).toDouble(),
      radius: (json['radius'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'name': name,
      'coordinates': [latitude, longitude],
      'radius': radius,
    };
  }
}